react/week3/todolist/src/App.js

```diff
@@ -0,0 +1,157 @@
+import React, { Component } from 'react';
+import './App.css';
+import uuid from 'uuid';
+
+import Todos from './components/Todos';
+import Timer from './components/Timer';
+
+
+class App extends Component {
+
+  constructor(props) {
+    super(props);
+          this.state =  {
+                todos:''
+              }
+
+          this.marked = this.marked.bind(this);
+          this.deleted = this.deleted.bind(this);
+          this.edit = this.edit.bind(this);
+          this.onSubmit = this.onSubmit.bind(this)
+          this.handleInputChange = this.handleInputChange.bind(this);
+        };
+   
+
+  componentDidMount =() =>{
+
+    const url = "https://gist.githubusercontent.com/benna100/391eee7a119b50bd2c5960ab51622532/raw"
+    fetch(url)
+    .then(response => response.json())
+    .then(data => {
+     data.forEach(element => {
+        const todo = {
+            id : element.id ,
+            description: element.description,
+            deadline: element.deadline,
+            checked: false,
+        };
+          this.setState({
+            todos: [...this.state.todos, todo]
+          })
+          
+         })  
+       })
+    };
+
+    onSubmit(e){
+      e.preventDefault();
+      const newItem = {
+        id : uuid(4),
+        description: this.state.description,
+        deadline: this.state.deadline,
+        checked: false,
+      };
+
+        this.setState({
+          todos: [...this.state.todos, newItem]
+        })
+        // e.target.reset()
+    }
+
+    handleInputChange(e) {
+      e.stopPropagation();
+        this.setState({
+          [e.target.name]: e.target.value,
+        }); 
+    }
+
+  marked =(id) =>{
+    this.setState({ todos : this.state.todos.map( todo => {
+        if (todo.id === id) {
+          todo.checked = !todo.checked
+        }
+        return todo
+      })});
+  }
+
+
+  edit(id , e ){   
+    const todosIndex= this.state.todos.findIndex(todo => todo.id ===id)
+    const todo = Object.assign({},this.state.todos[todosIndex]);
+    todo.description =this.state.description;
+    todo.deadline =this.state.deadline;
+    const todos = Object.assign([], this.state.todos);
+    todos[todosIndex] = todo;
+    this.setState({
+      todos: todos
+    })
```

Comment:
Another way determine which item to update is by comparing the updated todo id with corresponded todo id in the state, fx: 
```suggestion
 this.setState(prevState => {
            const newList = prevState.todos.map(todo => {
                if (id === todo.id) {
                    return {
                        ...todo,
                        description: prevState.description,
                        deadline: prevState.deadline,
                    };
                }
                return todo;
            });
            return { todos: newList };
        });
```
This way you don't have to worry about indexing.